New Env Vars Doc

paolaperaza opened the pull request:
Starting a draft PR for my changes to a new "Environment Variables" doc on Astronomer.

Addresses: https://github.com/astronomer/issues/issues/149

paolaperaza commented:
Alright, made some progress here! @vparekh94 @schnie Would love if you two could take a look at this. I tried to make it as detailed as possible to minimize open questions, but would love your feedback on accuracy, themes that might be missing entirely, etc.

Note: I know the "secret" feature isn't available until 0.16 and this is the 0.15 file - for simplicity I just threw it in but will take out later and contribute it up to the 0.16 docs folder that's in the works.

cc: @vishwas-astro

schnie commented on v0.15/environment-variables.md:
```diff
@@ -0,0 +1,235 @@
+
+## How Environment Variables are Stored on Astronomer
+
+All values for Environment Variables on Astronomer, whether added via your `Dockerfile` or via the Astronomer UI, are stored as a [Kubernetes Secret](https://kubernetes.io/docs/concepts/configuration/secret/), which is encrypted at REST and mounted into your Deployment's Airflow pods (Scheduler, Webserver, Worker(s)) as soon as they're set or changed.
```
To clarify, "encryption at rest" is referring to the difference between storing encrypted data on some sort of drive, vs data in transit across the internet, which is handled differently (tls).

paolaperaza replied:
Whoops, good call out. Adjusted.

paolaperaza commented on v0.15/environment-variables.md:
````diff
@@ -0,0 +1,235 @@
+If you run `ls` following this command, you'll see a list of running files:
+
+```
+bash-5.0$ ls
+Dockerfile             airflow.cfg            airflow_settings.yaml  dags                   include                logs                   packages.txt           plugins                requirements.txt       unittests.cfg
+```
````
Ah, great - thanks! Adjusted.

paolaperaza commented on v0.15/environment-variables.md:
```diff
@@ -0,0 +1,235 @@
+
+## How Environment Variables are Stored on Astronomer
+
+All values for Environment Variables on Astronomer, whether added via your `Dockerfile` or via the Astronomer UI, are stored as a [Kubernetes Secret](https://kubernetes.io/docs/concepts/configuration/secret/), which is encrypted at REST and mounted into your Deployment's Airflow pods (Scheduler, Webserver, Worker(s)) as soon as they're set or changed.
```
Ah, should've put that together. Got it - changed!

paolaperaza commented on v0.15/environment-variables.md:
```diff
@@ -0,0 +1,235 @@
+
+All values for Environment Variables on Astronomer, whether added via your `Dockerfile` or via the Astronomer UI, are stored as a [Kubernetes Secret](https://kubernetes.io/docs/concepts/configuration/secret/), which is encrypted at REST and mounted into your Deployment's Airflow pods (Scheduler, Webserver, Worker(s)) as soon as they're set or changed.
+
+Environment Variables are _not_ stored in Airflow's Metadata Database, are _not_ stored in our Platform Database and are _not_ made available in Astronomer's Houston API.
```
Didn't realize this, but makes total sense - got it. I presume the "Key" then would still be available via Houston, even though the value itself is `null`?

paolaperaza commented:
Ah, thank _you_ @schnie ! Just updated language based on your comments/corrections. I also added this note to the `Dockerfile` section:

> **Note:** Given that this file will be committed upstream, we strongly recommend witholding Environment Variables containing sensitive credentials from your `Dockerfile` and instead inserting them via your `.env` file locally (while adding the file to your `.gitignore`) or setting them as 'secret' via the Astronomer UI, as described in a dedicated section below.

There's a note in the `.env` section to add that file to `.gitignore` and then there's the whole section about setting them as secret in the UI. I'm thinking that's enough? Could still consider adding a section dedicated to answering the question of "When do I use which", if you guys think that'd be better.

vparekh94 commented on v0.15/environment-variables.md:
```diff
@@ -0,0 +1,240 @@
+
+> **Note:** As noted above, Workspace roles and permissions apply to actions in the `Variables` tab. For a full breakdown of permissions for each role, reference Astronomer's ["Roles and Permissions" doc](https://www.astronomer.io/docs/rbac/).
+
+### Precedence amongst 3 Methods
```
:heart_eyes:

paolaperaza replied:
:)

vparekh94 commented on v0.15/environment-variables.md:
```diff
@@ -0,0 +1,240 @@
+
+> **Note:** Input for all configurations officially supported by Airflow are pre-templated, but you're free to specify your own values.
+
+#### Mark Environment Variables as "Secret"
```
Is there an endpoint we expose to do this? I'd imagine someone who is trying to automate their deployment creation + readiness will want a way to _automatically_ add something and mark it as a secret.

paolaperaza replied:
Hmm good call - assumedly yes, though not sure if it's still just `updateDeployment`? Definitely happy to add instructions on that, either here or in the Houston API doc. Digging into this PR but not sure - https://github.com/astronomer/astro-ui/pull/523

@schnie Do you know?

schnie replied:
Yep its available in the API. Anything the UI or CLI can do will be available via direct API request as well.

It used to be the `updateDepoyment` mutation, but I believe it is now the `updateDeploymentVariables` mutation. It's new and dedicated to the env var purposes.

We also deprecated the old way of setting env vars, so that won't work in new deployments - only the new mutation will work.

We may want to start documenting this type of change somehow too.

vparekh94 commented:
@paolaperaza I don't have any examples; but I think the customers will be able to come up with those themselves. Think this is good to ship. Great work :)

paolaperaza commented:
Thank you for your v positive feedback, @vparekh94! I'll give it a couple more reads and merge for now so we can ship alongside 0.16 docs
